src/data/real/irrational.lean

```diff
@@ -257,4 +257,43 @@ open irrational
 @[simp] theorem irrational_inv_iff : irrational x⁻¹ ↔ irrational x :=
 ⟨of_inv, irrational.inv⟩
 
+/-!
+### Lemmas about Irrationals between Reals
+-/
+
+/-- There exists an irrational number between any two distinct rational numbers.-/
+theorem exists_irrational_btwn_rats {x y : ℚ} (h : x < y) :
+∃ r, irrational r ∧ (x : ℝ) < r ∧ r < y:=
+begin
+  refine ⟨x +((y - x) : ℚ) * (real.sqrt 2)⁻¹, _, _, _⟩,
+  { refine irrational.rat_add _
+      (irrational.rat_mul
+        (irrational_inv_iff.mpr irrational_sqrt_two) _),
+    exact_mod_cast (ne_of_gt (sub_pos_of_lt h)) },
+  { suffices : 0 < ↑(y - x) * (real.sqrt 2)⁻¹,
+    { simpa using h, },
+    refine (zero_lt_mul_right _).mpr _,
+    apply inv_pos.mpr,
+    norm_num,
+    exact_mod_cast (sub_pos_of_lt h) },
+  { apply (sub_lt_sub_iff_right (x :ℝ)).mp,
+    rw [add_sub_cancel', ←cast_sub],
+    refine (mul_lt_iff_lt_one_right _).mpr
+      (inv_lt_one _),
+    exact_mod_cast  (sub_pos_of_lt h),
+    refine (real.lt_sqrt zero_le_one (zero_le_bit0.mpr zero_le_one)).mpr _,
+    rw one_pow,
+    exact one_lt_two }
+end
```

Comment:
It might make sense to split this in three:
```suggestion
/-- An irrational number between any two distinct rational numbers.-/
def irrational_btwn_rats {x y : ℚ} (h : x < y) : { r : ℝ // irrational r } :=
⟨x +((y - x) : ℚ) * (real.sqrt 2)⁻¹, begin
  refine irrational.rat_add _
    (irrational.rat_mul
      (irrational_inv_iff.mpr irrational_sqrt_two) _),
  exact_mod_cast (ne_of_gt (sub_pos_of_lt h)) end⟩

lemma le_irrational_btwn_rats {x y : ℚ} (h : x < y) : x < (irrational_btwn_rats h : ℝ) :=
begin
  suffices : 0 < ↑(y - x) * (real.sqrt 2)⁻¹,
  { simpa using h, },
  refine (zero_lt_mul_right _).mpr _,
  apply inv_pos.mpr,
  norm_num,
  exact_mod_cast (sub_pos_of_lt h)
end

lemma irrational_btwn_rats_le {x y : ℚ} (h : x < y) : (irrational_btwn_rats h : ℝ) < y  :=
begin
  apply (sub_lt_sub_iff_right (x :ℝ)).mp,
  rw [add_sub_cancel', ←cast_sub],
  refine (mul_lt_iff_lt_one_right _).mpr (inv_lt_one _),
  exact_mod_cast  (sub_pos_of_lt h),
  refine (real.lt_sqrt zero_le_one (zero_le_bit0.mpr zero_le_one)).mpr _,
  rw one_pow,
  exact one_lt_tw
end
```
From which you can recover
```lean
theorem exists_irrational_btwn_rats {x y : ℚ} (h : x < y) :
   ∃ r, irrational r ∧ (x : ℝ) < r ∧ r < y :=
⟨irrational_btwn_rats h, subtype.prop _, le_irrational_btwn_rats h, irrational_btwn_rats_le h⟩,

Reply:
Wow thats pretty slick will do that, the dependent def is new to me.